tools/istio-docker.mk

```diff
@@ -74,13 +75,19 @@ DOCKER_FILES_FROM_SOURCE:=pilot/docker/prepare_proxy.sh docker/ca-certificates.t
 $(foreach FILE,$(DOCKER_FILES_FROM_SOURCE), \
         $(eval $(ISTIO_DOCKER)/$(notdir $(FILE)): $(FILE) | $(ISTIO_DOCKER); cp $(FILE) $$(@D)))
 
+# Copy the envoy images, but use standard file names (without the version) in docker
+$(ISTIO_DOCKER)/envoy: ${ISTIO_ENVOY_RELEASE_PATH} | $(ISTIO_DOCKER); cp ${ISTIO_ENVOY_RELEASE_PATH} $(ISTIO_DOCKER)/envoy
```

Comment:
How important is it to be able to specify/customize the envoy paths?  If envoy and envoy-debug was stored in ${ISTIO_OUT} rather than go/out then you wouldn't need these custom rules and could instead add "envoy" and "envoy-debug" to the list for DOCKER_FILES_FROM_ISTIO_OUT.  Actually, you'd also have to change this in the top-level Makefile to make knows how to resolve the envoy-debug:
<pre>
- ${ISTIO_OUT}/envoy: init
+ ${ISTIO_OUT}/envoy ${ISTIO_OUT}/envoy-debug: init
</pre>

The main problem with my suggestion is that ${ISTIO_OUT}/envoy is currently being used for some tests, so you'd have to do some seek-and-update to change such uses of "envoy" to "envoy-debug".  The other tradeoff of my suggestion is that is someone changes their cross-compile target then it'll re-download envoy for each, but in the long run maybe this is ideal for if/when istio gets to the point of compiling proxy for other platforms.

Reply:
Maybe envoy_debug would be better placed under OUT/$ARCH/debug ? 

In general, the current docker building is extremely inefficient - all binaries and files are passed to the docker builder ( copied ), when for most images only a tiny binary is needed.
Placing even more large binaries in the single common dir will make it worse.

Reply:
Good suggestions. I've changed the directories to be placed in ${OUT_DIR}/$[OS}_${ARCH}/<release/debug>